### ISSUE TITLE
Incorrect unreachable code detection when using typing.TYPE_CHECKING

### ISSUE THREAD

**Issue by harrismirza** [microsoft/pyright#8585]
**Describe the bug**
When using the typing.TYPE_CHECKING variable in a condition in code, anything in the else branch will be marked as unreachable. This is correct in the sense that Pyright can never reach that code, but from a user perspective, this is incorrect as they won't be running as a type checker.

Pyright should use the TYPE_CHECKING = True branch for checking types but mark the other branch as reachable.

**Code or Screenshots**
A small example is the following:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from datetime import datetime
else:
    from datetime import timedelta as datetime

print(datetime)
```

This shows the else branch as unreachable/

**VS Code extension or command-line**
This was first noticed via the Pylance extension (v2024.3.2), but also can be seen on the Pyright playground.

**Comment by harrismirza** [comment]
I have also created a unit test here to test for this behaviour.

I'm happy to take a crack at solving it as well, I should have some time this weekend

https://github.com/microsoft/pyright/commit/10ebc4efd26ccb191d59fec52ff75039c8bebd2c

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug. The behavior of the `TYPE_CHECKING` symbol is mandated by the Python typing spec. Type checkers are supposed to treat it as always `True` when performing type checking. You can think of `TYPE_CHECKING` as a way of lying to the type checker about what will be executed at runtime. This is useful in certain circumstances, but it should be used with care because the type checker will be analyzing code that's different from the code the interpreter sees.

Also, it's important for type checkers to treat the `not TYPE_CHECKING` branch as unreachable from a code analysis standpoint. If pyright were to analyze both the `if` and `else` blocks in your code sample, the symbol `datetime` would have conflicting definitions, which would lead to downstream type violations. That's not what users want or expect.

**Comment by harrismirza** [comment]
I agree with your points about the typechecking but pyright should at least silence the unreachable code warning on the output as the code is definitely reachable and will in fact be executed every time the code is run outside of a type checker.

**Comment by erictraut** [comment]
Pyright never emits warnings or errors for unreachable code blocks. It does emit a tagged hint if the client indicates that it supports tagged hints. This tells the editor to display the text in a subtle gray color to inform you that the region is not analyzed and, from the perspective of the type checker, is unreachable. 

If you don't want pyright to emit tagged hints, you can set `pyright.disableTaggedHints` to false in your LSP settings.
